### ISSUE TITLE
Convert **fast** reciprocal space and grinder to openmp

### ISSUE THREAD

**Issue by stuwilkins** [scikit-beam/scikit-beam#382]
Leave goodby ``pthreads`` and hello ``OpenMP``

**Comment by ericdill** [comment]
I am :+1: on converting to using OpenMP !  [Cython.parallel uses OpenMP for parallelization] (http://docs.cython.org/src/userguide/parallelism.html).  I would be very appreciative if we could seriously consider converting the ctrans.c into Cython for maintainability reasons.  The core devs of Python have suggested any C extensions should be written with a wrapping library like Cython or cffi since those libraries handle the c API compatibility between various versions of Python.

**Comment by stuwilkins** [comment]
Agreed! Using cpython as the “glue” between a ``c`` function call and python is better than the boiler plate to use the capi. However, I have reservation about **all** being in cython, and would prefer a c library with defined function calls, called by cython unless the underlying function is simple.

**Comment by ericdill** [comment]
So here's a question... As it currently stands, the fast gridder and fast q conversion code cannot be compiled on windows.  Should we plan on making this extension portable to windows at **any** point?

(windows has no pthread, see [here] (http://stackoverflow.com/questions/2150938/can-i-get-cs-pthread-h-to-compile-in-windows))

**Comment by licode** [comment]
OpenMP is a good option, and should be available for different systems.

**Comment by stuwilkins** [comment]
Thanks Li Li.

Yes if you notice I had already added an issue to do exactly this for the reasons you quite correctly point out. I will get to this, once the Beamline starts up.

There is also a compiler option to go to a single threw solution which would allow compilation on Windows however is anyone really using Windows for multi core data processing anyway???

Best wishes,
Stuart


On Jan 22, 2016, at 1:58 PM, Li Li <notifications@github.com<mailto:notifications@github.com>> wrote:


OpenMP is a good option, and should be available for different systems.

**Comment by ericdill** [comment]
Haha yes, I am not particularly concerned about using Windows for multi-core processing of gridding/q-conversion.  I see what you are saying about the USE_THREADS thing.  I suppose I could just change the setup.py from 

```python
def c_ext():
    return [Extension('skbeam.ext.ctrans', ['src/ctrans.c'],
                      define_macros=[('USE_THREADS', None)])]
```

to something like

```python
def c_ext():
    define_macros = []
    if os.name == 'posix':
        define_macros.append(('USE_THREADS', None))
    return [Extension('skbeam.ext.ctrans', ['src/ctrans.c'],
                      define_macros=define_macros)]
```

Also, can someone who has a mac report the output of `python -c "import os; print(os.name)"`

**Comment by sameera2004** [comment]
@ericdill output of `python -c "import os; print(os.name)"]` for mac
is `posix`

**Comment by ericdill** [comment]
Fantastic, thanks!

**Comment by licode** [comment]
I am implementing openmp on another c/c++ code. So I can also help on this later.
